### PR TITLE
override instead of wrap TesserocrRecognize in other processors

### LINKED ISSUE #190: Use ocrd-tesserocr with caching for Processing Server not possible
In the [processing server pull request](https://github.com/OCR-D/core/pull/974) we want to use caching. Therefore an [instance of a processor without a workspace](https://github.com/OCR-D/core/blob/fd5effb967bad1f2ff860e3f5efd41997c77003d/ocrd/ocrd/processor/helpers.py#L283-L299
) is created. Later when the user requests a processor-run we want to [set the workspace into that cache-instance](https://github.com/OCR-D/core/blob/fd5effb967bad1f2ff860e3f5efd41997c77003d/ocrd/ocrd/processor/helpers.py#L313-L318).

The problem seems to be that ocrd-tesserocr processors delegates the processing internally to the processor TesserocrRecognize`:
https://github.com/OCR-D/ocrd_tesserocr/blob/c10f94d2b36815314687a216495e8602b4a16bb0/ocrd_tesserocr/segment_region.py#L13-L32

The workspace is transferred to this class on creation. If the workspace is (re-)set afterwards it does not get delegated to the created instance of `TesserocrRecognize`

### PR DESCRIPTION
**Pull request by bertsky**
fixes #190 

Perhaps we should also add a test case for the workspace mechanics. I have a feeling this should have failed long ago even without the Processing Server's instance caching.

### CODE REVIEW

**Comment by bertsky** [comment]
> Perhaps we should also add a test case for the workspace mechanics. I have a feeling this should have failed long ago even without the Processing Server's instance caching.

Ah, it actually [did](https://app.circleci.com/pipelines/github/OCR-D/ocrd_tesserocr/485/workflows/61c6442c-052f-4037-a550-c18033255827/jobs/635)! (We just had no event on this repo trigger the CI since the last get_processor / run_processor changes in core.)

**Comment by bertsky** [comment]
Turns out I had to do a lot more to fix the CI:
- build based on ocrd/core but with PIP and PYTHON variables did not work (anymore?) and is not useful: we want to test various Python versions natively (without the hassle of OS dependencies for them), so the base must be a matrix of Python CI images
- Python 3.6 now (being end-of-life) builds much faster if older Numpy/OpenCV binaries get installed beforehand
- the PPA for Tesseract does not work (and is not necessary) for the most up-to-date Python versions (i.e. OS versions), so this step (including the related `apt-get update`) must be optional
- the PPA version of Tesseract has its tessdata directory (now identical to our ocrd-tesserocr moduledir, where resmgr wants to install additional models into) owned by root, so this must become writable for others

**Comment by bertsky** [comment]
Ok, I ran into https://github.com/OCR-D/core/issues/998. Now, I could wait for the fix to be merged, or avoid the `kwargs.setdefault('ocrd_tool', ...)` pattern. But I don't want either, so let's do a workaround for now.

**Review by joschrew — APPROVED** [review]
I have tested this code together with the [processing-server PR](https://github.com/OCR-D/core/pull/974) and everything works now. So this PR seems to solve my problems regarding the tesser-ocrd-processors.
Additionally I have briefly gone through the python code changes (I ignored all ci-cd changes). I didn't find anything which I could remark here (except my question).

File: ocrd_tesserocr/segment_region.py
Comment (bertsky) [reply ocrd_tesserocr/segment_region.py:18]: Because there are different contexts for processor instances (depending on CLI arguments):
- instances that will be passed to `run_processor` (which ultimately runs `process`, ideally repeatedly)
- instances that exit immediately and print/dump some info

To differentiate this, by the CLI and API we can only check whether or not an `output_file_grp` (or `parameter` or `workspace`) kwarg exists.

**Comment by bertsky** [comment]
@kba if you make a release out of https://github.com/OCR-D/core/pull/999 today, I'll revert [b357217](https://github.com/OCR-D/ocrd_tesserocr/pull/191/commits/b357217e6fd5610d8307cc6ca67f8ab55a9fb3ed) here and require that new core version.

**Comment by kba** [comment]
> 
@kba if you make a release out of OCR-D/core#999 today, I'll revert b357217 here and require that new core version.

v2.47.0 has been released now.

**Comment by stweil** [comment]
> v2.47.0 has been released now.

Too early, because we need https://github.com/OCR-D/core/pull/1004.

**Comment by bertsky** [comment]
> v2.47.0 has been released now.

Thanks! I'll adapt...

> Too early, because we need [OCR-D/core#1004](https://github.com/OCR-D/core/pull/1004).

This is not about Docker, but Python API.

**Comment by kba** [comment]
> > v2.47.0 has been released now.
> 
> Too early, because we need [OCR-D/core#1004](https://github.com/OCR-D/core/pull/1004).

Hotfix incoming.

**Comment by bertsky** [comment]
> > > Ok, I ran into [OCR-D/core#998](https://github.com/OCR-D/core/issues/998). Now, I could wait for the fix to be merged, or avoid the `kwargs.setdefault('ocrd_tool', ...)` pattern. But I don't want either, so let's do a workaround for now.

> > v2.47.0 has been released now.
> 
> Thanks! I'll adapt...

Odd. CI failure did come back:
```
  File "/home/circleci/project/ocrd_tesserocr/cli.py", line 23, in ocrd_tesserocr_segment_region
    return ocrd_cli_wrap_processor(TesserocrSegmentRegion, *args, **kwargs)
  File "/home/circleci/.pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/decorators/__init__.py", line 116, in ocrd_cli_wrap_processor
    run_processor(processorClass, mets_url=mets, workspace=workspace, **kwargs)
  File "/home/circleci/.pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 95, in run_processor
    instance_caching=instance_caching
  File "/home/circleci/.pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 329, in get_processor
    parameter=parameter
  File "/home/circleci/project/ocrd_tesserocr/segment_region.py", line 17, in __init__
    self.parameter['overwrite_segments'] = self.parameter['overwrite_regions']
KeyError: 'overwrite_regions'
```

That's despite `ocrd==2.47.1`. Where does the empty `ocrd_tool` come from???

**Comment by bertsky** [comment]
> That's despite `ocrd==2.47.1`. Where does the empty `ocrd_tool` come from???

Aha! It turns out that my https://github.com/OCR-D/core/pull/999 was premature: we _also_ pass an unnecessary `ocrd_tool` in other places:
- run_processor kwarg: https://github.com/OCR-D/core/blob/cff8ddaefec72451795a5e0b4bdb7efc18223d99/ocrd/ocrd/processor/helpers.py#L38
- run_processor calling get_processor: https://github.com/OCR-D/core/blob/cff8ddaefec72451795a5e0b4bdb7efc18223d99/ocrd/ocrd/processor/helpers.py#L91
- get_processor kwarg: https://github.com/OCR-D/core/blob/cff8ddaefec72451795a5e0b4bdb7efc18223d99/ocrd/ocrd/processor/helpers.py#L306
- get_processor calling constructor: https://github.com/OCR-D/core/blob/cff8ddaefec72451795a5e0b4bdb7efc18223d99/ocrd/ocrd/processor/helpers.py#L325

Now, the big story here is: **none of these places is needed** (or used) – neither currently, nor for OCR-D/core#974 nor OCR-D/core#884!

**Comment by joschrew** [comment]
Other questions (hopefully it is ok to ask it here):
- What is the reason to do this `kwargs.setdefault('ocrd_tool', OCRD_TOOL['tools'][TOOL])` instead of doing this: `kwargs['ocrd_tool'] = OCRD_TOOL['tools'][TOOL]` (this is the way it was done before). Why shouldn't the ocrd_tool always be set in the constructor?
- I tried to get the idea behind the ocrd-tool-setting. Is it right that a processor-constructor is supposed to set the ocrd_tool always before calling the super-constructor? I have only viewed a few example processors and I am wondering if this (setting ocrd-tool in a processors __init__() before calling the super constructor) is something like a rule (In Java you would probably use something like an abstract method in the processor-base-class?)?

**Comment by bertsky** [comment]
> * What is the reason to do this `kwargs.setdefault('ocrd_tool', OCRD_TOOL['tools'][TOOL])` instead of doing this: `kwargs['ocrd_tool'] = OCRD_TOOL['tools'][TOOL]` (this is the way it was done before). Why shouldn't the ocrd_tool always be set in the constructor?

The reason is that here in ocrd_tesserocr many processors are basically just re-parameterizations (simplifications) of the "goliath" ocrd-tesserocr-recognize. This helps the user cope with complexity (and avoids code duplication while sustaining the older single-step processors)  – see README. 

To implement this, you have to know that ocrd.Processor does parameter parsing/instantiation/validation on the constructor:
https://github.com/OCR-D/core/blob/cff8ddaefec72451795a5e0b4bdb7efc18223d99/ocrd/ocrd/processor/base.py#L146-L150

So you need to
1. enter with the (possibly empty) params for the wrapper processor, 
2. let them validate (instantiate) against the wrapper processor's tool json,
3. translate them to the TesserocrRecognize's tool json,
4. re-validate (instantiate) against that tool json.

Now, in the old pattern, where every ocrd.Processor's subclass constructor just overwrites the kwarg for `ocrd_tool` with its own local tool json (from pkg_resources), we cannot bring TesserocrRecognize to accept any foreign tool json. Hence the `setdefault` pattern, which IMO is also better in general.

> * I tried to get the idea behind the ocrd-tool-setting. Is it right that a processor-constructor is supposed to set the ocrd_tool always before calling the super-constructor?

Yes, it must. Otherwise, the parameter validator in the superclass won't see the actual cmdline values. Also, the non-processing contexts (help, dump json, show resources) in the superclass constructor need to see the actual tool json.

> I have only viewed a few example processors and I am wondering if this (setting ocrd-tool in a processors `__init__()` before calling the super constructor) is something like a rule

We could of course change the API in core to separate the parameter instantiation, but that would entail changing all processors (lots of diverse codebases with various maintainers). We have to do that for the new `process_page` [API](https://github.com/OCR-D/core/issues/322) soon anyway, so it's a good time to think about alternatives...

> (In Java you would probably use something like an abstract method in the processor-base-class?)?

Yes. We could define a method (say) `init` which (in the subclasses) loads the tool json, and gets called (by the superclass) before most of the work done in the current superclass constructor.

For the processor delegation pattern then, since in the [new API](https://github.com/OCR-D/core/issues/322), we will also have a `setup` for the processing context of a processor instance, we could do parameter re-instantiation for the new superclass there. So the subclasses would simply have to call `super().setup()` and then do their own ParameterValidator.validate.

**Comment by mikegerber** [comment]
With this PR (and ocrd 2.48.0) I seem to have problems relating to $TESSDATA_PREFIX:
```
$ ocrd-tesserocr-segment-region -I OCR-D-IMG-BINPAGE-sauvola -O OCR-D-TEST-TEST-TEST
18:37:10.199 ERROR ocrd.processor.helpers.run_processor - Failure in processor 'ocrd-tesserocr-segment-region'
Traceback (most recent call last):
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 128, in run_processor
    processor.process()
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd_tesserocr/recognize.py", line 347, in process
    oem=getattr(OEM, self.parameter['oem']))
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd_tesserocr/recognize.py", line 95, in InitFull
    super().InitFull(path=self.path, lang=self.lang, oem=self.oem, variables=self.parameters)
  File "tesserocr.pyx", line 1457, in tesserocr.PyTessBaseAPI.InitFull
  File "tesserocr.pyx", line 1245, in tesserocr.PyTessBaseAPI._init_api
RuntimeError: Failed to init API, possibly an invalid tessdata path:
Traceback (most recent call last):
  File "/usr/local/share/pyenv/versions/3.7.16/bin/ocrd-tesserocr-segment-region", line 11, in <module>
    load_entry_point('ocrd-tesserocr==0.16.0', 'console_scripts', 'ocrd-tesserocr-segment-region')()
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd_tesserocr/cli.py", line 23, in ocrd_tesserocr_segment_region
    return ocrd_cli_wrap_processor(TesserocrSegmentRegion, *args, **kwargs)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/decorators/__init__.py", line 116, in ocrd_cli_wrap_processor
    run_processor(processorClass, mets_url=mets, workspace=workspace, **kwargs)
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 131, in run_processor
    raise err
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 128, in run_processor
    processor.process()
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd_tesserocr/recognize.py", line 347, in process
    oem=getattr(OEM, self.parameter['oem']))
  File "/usr/local/share/pyenv/versions/3.7.16/lib/python3.7/site-packages/ocrd_tesserocr/recognize.py", line 95, in InitFull
    super().InitFull(path=self.path, lang=self.lang, oem=self.oem, variables=self.parameters)
  File "tesserocr.pyx", line 1457, in tesserocr.PyTessBaseAPI.InitFull
  File "tesserocr.pyx", line 1245, in tesserocr.PyTessBaseAPI._init_api
RuntimeError: Failed to init API, possibly an invalid tessdata path:
$ echo $TESSDATA_PREFIX
/home/mike.gerber/.local/share/ocrd-resources/ocrd-tesserocr-recognize
$ ls $TESSDATA_PREFIX
Fraktur_GT4HistOCR.traineddata  deu.traineddata  frk.traineddata
```

**Comment by kba** [comment]
```
ls $TESSDATA_PREFIX
Fraktur_GT4HistOCR.traineddata  deu.traineddata  frk.traineddata
```

You need `osd.traineddata` I think. Can you try `ocrd resmgr download ocrd-tesserocr-recognize osd.traineddata` and check if that fixes it?

**Comment by mikegerber** [comment]
Ah. Never had it working on this system, and I wrongly assumed it was a new bug.

I copied everything from /usr/share/.../tessdata for now, that fixed `ocrd-tesserocr-segment-region`! This PR's `ocrd-tesserocr-recognize` also works for me.

**Comment by bertsky** [comment]
@kba this time, codecov broke – IIUC it says it now has 0% coverage, because I have added more processors to the test :roll_eyes: – do you have any idea what's going on?

**Comment by bertsky** [comment]
Understood. Thanks!

I'll merge and release then.

**Pull request merged**
